cmd/router/main.go

```diff
@@ -351,11 +365,43 @@ func main() {
 		os.Exit(1)
 	}
 
+	server := &http.Server{
+		Addr:    ":" + strconv.Itoa(constants.RouterPort),
+		Handler: nil, // default server mux
+		// https://medium.com/a-journey-with-go/go-understand-and-mitigate-slowloris-attack-711c1b1403f6
+		ReadHeaderTimeout: time.Minute,
+	}
 	http.HandleFunc("/", graphHandler)
+	http.HandleFunc(constants.RouterReadinessEndpoint, readyHandler)
 
-	err = http.ListenAndServe(":8080", nil)
-	if err != nil {
-		log.Error(err, "failed to listen on 8080")
+	go func() {
+		err = server.ListenAndServe()
+		if err != nil && !errors.Is(err, http.ErrServerClosed) {
+			log.Error(err, fmt.Sprintf("Failed to serve on address %v", server.Addr))
+			os.Exit(1)
+		}
+	}()
+
+	// Blocks until SIGTERM or SIGINT is received
+	handleSignals(server)
+}
+
+func handleSignals(server *http.Server) {
+	signalChan := make(chan os.Signal, 1)
+	signal.Notify(signalChan, os.Interrupt, syscall.SIGTERM)
+
+	sig := <-signalChan
+	log.Info("Received shutdown signal", "signal", sig)
+	// Fail the readiness probe
+	isShuttingDown = true
+	log.Info("Sleeping %v to allow K8s propagation of non-ready state", drainSleepDuration)
+	// Sleep to give networking a little bit more time to remove the pod
+	// from its configuration and propagate that to all loadbalancers and nodes.
+	time.Sleep(drainSleepDuration)
+	// Shut down the server gracefully
+	if err := server.Shutdown(context.Background()); err != nil {
+		log.Error(err, "Failed to shutdown the server gracefully")
```

Comment:
If graceful shutdown fails, can we do force close something like this !


```
// Shut down the server forcefully if context deadline exceeded.
		if errors.Is(err, context.DeadlineExceeded) {
			log.Info("Shutting down the server forcefully ...")
			if err := server.Close(); err != nil {
				log.Error(err, "Failed to shutdown the server forcefully")
				os.Exit(1)
			}
			log.Info("Server forcefully shutdown")
		}
```

Reply:
We are using background context which will not timeout, so we don't need this logic.